### PR TITLE
feat: shared public context

### PR DESCRIPTION
**Pull request by Asuka109**
# PR Details

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/builder/builder-shared/src/createContext.ts [hunk 1]
```diff
@@ -42,36 +40,49 @@ export function createContextByConfig(
...
-    set(target, prop: keyof BuilderContext) {
+    set(_target, prop: keyof SharedBuilderContext) {
       logger.error(
         `Context is readonly, you can not assign to the "context.${prop}" prop.`,
       );
       return true;
     },
-  });
-}
+  }) as any;
```
Comment (chenjiahan) [inline packages/builder/builder-shared/src/createContext.ts:87]: 为啥有个 any..

File: packages/builder/builder-shared/src/createContext.ts [hunk 2]
```diff
@@ -42,36 +40,49 @@ export function createContextByConfig(
...
+  'srcPath',
+  'rootPath',
+  'distPath',
+  'devServer',
+  'framework',
+  'cachePath',
+  'configPath',
+  'tsconfigPath',
+  'bundlerType',
+  'hooks',
```
Comment (chenjiahan) [inline packages/builder/builder-shared/src/createContext.ts:55]: 这里已经有 hooks 了，为啥其他地方还要通过 additionalKeys 再传一个 `['hooks']`

Comment (Asuka109) [reply packages/builder/builder-shared/src/createContext.ts:55]: 确实不太合理，我重新组织下

File: packages/builder/builder-shared/src/createContext.ts (hunk 1 above)
Comment (Asuka109) [reply packages/builder/builder-shared/src/createContext.ts:87]: 我改下

**Review by chenjiahan — DISMISSED** [review]

**Pull request closed without merging**